### ISSUE TITLE
Completion dialog sometimes not shown when filenames with space exist (prompt_toolkit)

### ISSUE THREAD

**Issue by ju-kreber** [xonsh/xonsh#3843]
## xonfig

<details>

```
$ xonfig
+------------------+-----------------+
| xonsh            | 0.9.22          |
| Python           | 3.7.7           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.5           |
| shell type       | prompt_toolkit  |
| pygments         | 2.6.1           |
| on posix         | True            |
| on linux         | True            |
| distro           | unknown         |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+

```

</details>

## Expected Behavior

I'm in a directory with multiple files where at least one of them contains a space.
I start entering a command like `file ` and press tab.
A dialog should open, listing all the files in the directory (quoted) for completion.

## Current Behavior

_Sometimes_ no dialog with completion suggestions is shown.
Only the first completion is chosen and pressing tab iterates through all completions (possibly some fallback behavior?).

_Edit:_ Include a recording of the issue (with `ls` instead of `file`):
![xonsh_completion_bug](https://user-images.githubusercontent.com/8841729/95678438-e0f02300-0bcc-11eb-8961-36cc6b3f08e7.gif)

## Steps to Reproduce

When exactly this happens is really weird, but I found out a way to reproduce this issue (at least it works relatively consistently for me).
Create said directory with a bunch of files where at least one them them contains spaces.
Typing `file <tab>` now shows the undesired behavior.
**To fix it** simply complete something where suggestions do not contain spaces, like `file /<tab>` (usually) of simply `fi<tab>`. The dialogue is shown as expected and also the original `file <tab>` lists the quoted files as expected now.
**To break it** again just press enter a few times. This really did it for me. It doesn't matter if you execute a command or enter a completely empty line, after a few new lines, `file <tab>` behaves unexpectedly again.

I really have no idea what causes this and hope it's not limited to my case only. I tried to have a look at the internal workings of prompt_toolkit, but I'm too unfamiliar with it to figure something out, it seems :/

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by scopatz** [comment]
Thanks for reporting @ju-kreber! This may because `file` is a Python builtin, so just `file <TAB>` is ambiguous between Python and subprocess modes. Can you reproduce this with another command, such as `ls`?

**Comment by ju-kreber** [comment]
Hi, yes, it behaves the same for `ls`, `touch`, `head`, ... Sorry for the bad example!

**Comment by bobhy** [comment]
I can take a look at this, I've noticed some weirdness in both prompt toolkit completion menus (when there's only one row of matches) and in the file path completer when there are files with blanks in the name in the folder (whether or not they are displayed as completions).

@ju-kreber, there are a couple of known behaviors of prompt toolkit that might be involved here, please comment...

Prompt toolkit shell is not expected to display the completion menu if there is only one possible match when you hit TAB, it just completes that match on the command line.  It (often!) does this now, and I'd argue this is the way it should work.  But that doesn't sound like what you're describing, right?

It is also expected to automatically select the first completion in the menu, regardless of how many there are.  (This is a recent change in Xonsh).

Another known behavior is that prompt toolkit doesn't display as many rows of completion menu when the prompt line is at or near the bottom of the screen as when it's at the top.   In the worst case, when the prompt is right at the bottom of the screen and when there is only one row of completions to display, PTK will show the completions *on top of* the prompt rather than just below it.  Does this sound like what you're seeing?

So, could you please try this:

Make sure you have "many" files in the folder (enough to require multiple rows in the completion menu), then try your scenario with the prompt at the top of the screen (e.g, `$ clear`, then `$ ls <TAB>`) and at the bottom of the screen (`$ ll; ll; ll` to fill the screen, then `$ ls <TAB>`) and see if it behaves differently.  Also try your scenario with and without the file-with-embedded-blank present in the folder.   I have noticed that the file quoting works differently when the file is present, whether or not it is in the completion menu (e.g, if file is `'argle bargle'` but you do a completion like `$ ls b<TAB>`)

Finally, are you running on Windows?  With or without $FORCE_POSIX_PATHS == True?  On Windows without $FORCE_POSIX_PATHS, I've noticed doubled backslashes on folder names, (e.g folder name `foo/` is displayed as `foo\\` instead of `foo\`), but only when there is a file with embedded blank somewhere in the folder.

**Comment by ju-kreber** [comment]
Hi @bobhy, thank you so much for investigating!

I did not notice, but it seems like you're right, it really depends on whether I'm at the bottom of the screen or not. I don't think I require long names or many files for this to happen. Apparently, the behavior is the following: Usually when at the bottom and opening a completion dialogue, xonsh makes space for it by a few line feeds. For some reason, when spaces are contained in the file names, this fails.

I also did a quick screen recording to show the exact behavior:
![xonsh_completion_bug](https://user-images.githubusercontent.com/8841729/95678438-e0f02300-0bcc-11eb-8961-36cc6b3f08e7.gif)

For your additional questions, yes, I'm using the multi style and I'm on Linux (Mint 18).

**Comment by bobhy** [comment]
Hi, @ju-kreber thanks for the very clear GIF.  A picture is worth a thousand words, indeed!
That looks pretty much like the issue I am working on.  I'm convinced it's prompt-toolkit "behavior" (design, not bug) that doesn't play well with completion menus displayed at the bottom of the screen. 

I cannot explain why the presence of a file with embedded space should matter.  But 3 files in a completion menu is a magic number to PTK.  Could you try your repro with only files 'a', 'b' and 'with spaces'?  That might change what you see... 

I have never seen PKT completely cut off the whole completion menu at the bottom of the screen like that, but I believe it could. 
 Perhaps prompt-toolkit is not getting the right answer when it queries the terminal for number of rows?  Does `bash echo $LINES $COLUMNS` report correct numbers?  Are you having any other display issues with `more` or `vim` not showing the status line at the bottom of screen?  These use the alternate buffer but depend on the same VT100 sequences to determine screen size.

The easiest "workaround" I can suggest for now is set `$COMPLETIONS_DISPLAY = 'readline'` and don't worry about the menus.  I personally find this easier to use, FWIW.

But I do have a set of changes in prompt-toolkit and a small one in xonsh which handle the bottom-of-screen scenario better, though neither has been merged yet.  You're welcome to give it a try, if you have the time and patience for it.  Here's the sequence:

```
bobhy@PC:~/src-alt/xonsh$ python3 -m venv ~/venv/x1
bobhy@PC:~/src-alt/xonsh$ . ~/venv/x1/bin/activate

(x1) bobhy@PC:~$ mkdir src-alt
(x1) bobhy@PC:~$ cd src-alt

(x1) bobhy@PC:~/src-alt$ git clone https://github.com/xonsh/xonsh
(x1) bobhy@PC:~/src-alt$ git clone https://github.com/prompt-toolkit/python-prompt-toolkit

(x1) bobhy@PC:~/src-alt$ cd xonsh
(x1) bobhy@PC:~/src-alt/xonsh$ git pull origin pull/3874/head:completion_menu_rows
(x1) bobhy@PC:~/src-alt/xonsh$ git checkout completion_menu_rows
(x1) bobhy@PC:~/src-alt/xonsh$ pip install -r requirements/development.txt

(x1) bobhy@PC:~/src-alt/xonsh$ pip install -e .
Obtaining file:///home/bobhy/src-alt/xonsh
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Installing collected packages: xonsh
  Attempting uninstall: xonsh
    Found existing installation: xonsh 0.9.23
    Uninstalling xonsh-0.9.23:
      Successfully uninstalled xonsh-0.9.23
  Running setup.py develop for xonsh
Successfully installed xonsh

(x1) bobhy@PC:~/src-alt/xonsh$ cd ../python-prompt-toolkit/
(x1) bobhy@PC:~/src-alt/python-prompt-toolkit$ git pull origin pull/1271/head:completion_menu_rows
(x1) bobhy@PC:~/src-alt/python-prompt-toolkit$ git checkout completion_menu_rows

(x1) bobhy@PC:~/src-alt/python-prompt-toolkit$ pip install -e .
Obtaining file:///home/bobhy/src-alt/python-prompt-toolkit
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: wcwidth in /home/bobhy/venv/x1/lib/python3.8/site-packages (from prompt-toolkit==3.0.8) (0.2.5)
Installing collected packages: prompt-toolkit
  Attempting uninstall: prompt-toolkit
    Found existing installation: prompt-toolkit 3.0.8
    Uninstalling prompt-toolkit-3.0.8:
      Successfully uninstalled prompt-toolkit-3.0.8
  Running setup.py develop for prompt-toolkit
Successfully installed prompt-toolkit

(x1) bobhy@PC:~/src-alt/xonsh$ xonsh
```

**Comment by ju-kreber** [comment]
Alright, first off, having only 3 files (`a`, `b`, `with spaces`) does not change the behavior at all. What I noticed is that when making space for the completion window works, it seems to always make space for exactly 4 lines, no matter the number of possible completions. I don't know if that's intended or not..

In an interactive bash session, `echo $LINES $COLUMNS` outputs `32 238` which seems pretty reasonable for my guake terminal. `vim` and `nano` all work as expected as far as I can tell.

Finally, I tried out the steps you described and **it works perfectly!** So it seems that your fix works just as intended. I tried with few, with many files, with long and short names and I did not encounter any case where it broke. Interestingly, the space that is made for the completion window now also is exactly as much as required.
Thank you so much for fixing this!
